3.1

robynstuart opened the pull request:
### Brief description
This PR continues the modularization of FPsim by refactoring Education and Contraception into Starsim modules. Among other things, this will mean that other models can make use of them. It also removes the empowerment logic so that this can live independently in the empowerment repo. 

As part of removing the empowerment logic, this PR also removes the longitudinal circular buffer. We already have an  analyzer that takes snapshots of any metrics we want to keep track of over time, so it'd be easier to use this if we need to.

Fixes #594, #583, #586 

### Type(s) of change
- [ ] Bugfix
- [X] Refactor
- [ ] New feature
- [ ] Other

### Checklist
- My code follows the [style guide](https://github.com/amath-idm/styleguide)
  - [x] Yes
  - [ ] N/A
- I've commented my code
  - [x] Yes
  - [ ] N/A
- I've incremented the version number
  - [x] Yes
  - [ ] N/A
- I've updated the changelog
  - [x] Yes
  - [ ] N/A
- I've added tests and checked code coverage
  - [x] Yes
  - [ ] N/A

RHull-IDM commented on fpsim/methods.py:
```diff
@@ -122,58 +154,76 @@ def remove_method(self, method_label):
             force_choose=False,  # Whether to force non-users to choose a method
             method_weights=np.ones(self.n_methods),
-            max_dur=100*fpd.mpy,  # Maximum duration of use in months
+            max_dur=ss.years(100),  # Maximum duration of use in months
```
Comment specifies months but using years now

robynstuart commented on tests/benchmark.json:
```diff
@@ -1,15 +1,15 @@
-    "run": 6.718,
-    "postprocess": 0.03,
-    "total": 6.75
+    "initialize": 0.006,
+    "run": 15.999,
+    "postprocess": 0.049,
+    "total": 16.054
```
I believe this is because of switching some operations to CRN...

robynstuart replied:
@RHull-IDM would you mind running update_baseline to see what performance difference you see?

MObrien-IDM reviewed (commented):
I'd like to see a more descriptive note in the changelog since we are jumping to Version 3.1.0. Please expand and then I'll approve. Thanks!

robynstuart commented:
@MObrien-IDM I've added more detail to the changelog
@RHull-IDM if possible, would you mind reviewing again, as I've added some more changes to the parameter initialization. Thank you!

RHull-IDM reviewed (changes requested):
This is a great PR. I really like how you extended the pars concept I started implementing, and now that so much of the logic has been moved to connectors People.py is looking positively tidy by comparison. I didn't see anything critical in my review, but had a few comments for your consideration.

robynstuart commented on fpsim/defaults.py:
```diff
@@ -17,43 +17,10 @@
-#             shape = n if self.ncols is None else (n, self.ncols)
-#             arr = np.full(shape=shape, fill_value=vals, dtype=dtype)
-#         return arr
+valid_country_locs = ['senegal', 'kenya', 'ethiopia']
```
That's right, regions can be specified on their own, i.e. `sim1 = fp.Sim(location='addis_ababa')`

robynstuart commented on fpsim/locations/ethiopia/regions/data/age_pyramid.csv:
Oh interesting - had not thought about this, but I've added a comment to the issue [here](https://github.com/fpsim/fpsim/issues/584). That issue is actually about potentially moving the datafiles to a separate repo, but an alternative could be to streamline them and maintain them here.

robynstuart commented on fpsim/methods.py:
```diff
@@ -120,62 +192,68 @@ def add_method(self, method):
         self.methods[method.name] = method
 
     def remove_method(self, method_label):
+        errormsg = ('remove_method is not currently functional. See example in test_parameters.py if you want to run a '
```
Fixed

robynstuart commented on tests/benchmark.py:
```diff
@@ -30,12 +29,11 @@ def run_sim():
     to_profile = sc.objdict(
         run             = sim.run,
-        update          = ppl.update,           # 70% of sim.run() runtime is spent here
+        step            = ppl.step,           # 70% of sim.run() runtime is spent here
```
I've just removed these comments, as they aren't right anymore. I've added a comment noting that we should update this function once the Starsim v3 port is complete, as that will give us a lot more profiling functionality.

robynstuart commented on tests/test_analyzers.py:
```diff
@@ -87,14 +87,14 @@ def test_longitudinal():
-    assert max_age - min_age == pytest.approx(1 - 1/sim.fp_pars['tiperyear'], rel=1e-2), 'Expected age difference to be equal to 1 year minus the timestep size'
+    min_age = sim.analyzers.longitudinal_history.age[ss.uids(1), 0]
+    max_age = sim.analyzers.longitudinal_history.age[ss.uids(1), 1]
+    assert max_age - min_age == pytest.approx(1/sim.fp_pars['tiperyear'], rel=1e-2), 'Expected age difference to be equal to 1 year minus the timestep size'
```
I think this is ok now - I had originally changed the test because I had misunderstood what it was doing, but I've since changed it back so it's the same as you originally wrote it.

robynstuart commented on tests/test_sim.py:
```diff
@@ -50,9 +61,59 @@ def test_mid_choice():
+    assert sim3.fp_pars.postpartum_dur == 18, "FP par failed"
+
+    print('✓ (successfully created sims with different methods)')
+
```
Good idea - added!
